### PR TITLE
Made unhappiness effects moddable by adding a global uniques json; added revolts when < -20 happiness

### PR DESCRIPTION
**Pull request by xlenstra**
Allowing for more moddability.

### CODE REVIEW

**Comment by yairm210** [comment]
This is a good idea, but I'll raise you one: *conditional effects*
Each effect has a *list* of conditions 
That way, we can convert other civ-wide *and mayve city-wide effects* to this format

What I'm thinking of is Golden Age, but I'm sure there are other cases!

File: core/src/com/unciv/ui/tutorials/TutorialController.kt [hunk 1]
```diff
@@ -70,10 +71,20 @@ class TutorialController(screen: BaseScreen) {
...
+        }.map { tutorial ->
+            val lines = tutorial.value
+                
+            if (tutorial.key == "Unhappiness") {
+                for (unhappinessEffect in ruleset.unhappinessEffects.values.sortedByDescending { it.unhappiness }) {
+                    lines.add("\n${unhappinessEffect.toCivilopediaLines()}")
+                }
+            }
+            
+            CivilopediaTutorial(tutorial.key, lines)
```
Comment (SomeTroglodyte) [inline core/src/com/unciv/ui/tutorials/TutorialController.kt:86]: Nicely done.
Wasn't there another virtually generated page in there, to compare....? No, that was Religion, wasn't it?
And - "\n": Is that a FormattedLine collection or Strings? If the former, there is a "styled" empty line with a global-const controlled spacing (smaller than an entire line when I designed it) - if I remember correctly - thus mixing gives barely detectable line spacing inconsistencies.

**Comment by xlenstra** [comment]
> This is a good idea, but I'll raise you one: _conditional effects_ Each effect has a _list_ of conditions That way, we can convert other civ-wide _and mayve city-wide effects_ to this format
> 
> What I'm thinking of is Golden Age, but I'm sure there are other cases!

This sounds like a great idea!

**Comment by xlenstra** [comment]
> This is a good idea, but I'll raise you one: _conditional effects_ Each effect has a _list_ of conditions That way, we can convert other civ-wide _and mayve city-wide effects_ to this format
> 
> What I'm thinking of is Golden Age, but I'm sure there are other cases!

This sounds like a great idea!

**Comment by xlenstra** [comment]
> This is a good idea, but I'll raise you one: _conditional effects_ Each effect has a _list_ of conditions That way, we can convert other civ-wide _and mayve city-wide effects_ to this format
> 
> What I'm thinking of is Golden Age, but I'm sure there are other cases!

This sounds like a great idea!

**Comment by ravignir** [comment]
Add "Cannot settle new cities" if unhappiness is below -10.

**Comment by xlenstra** [comment]
> This is a good idea, but I'll raise you one: _conditional effects_ 
> Each effect has a _list_ of conditions 
> That way, we can convert other civ-wide _and maybe city-wide effects_ to this format
> 
> What I'm thinking of is Golden Age, but I'm sure there are other cases!

I suppose I can convert the changes made here to that format, though the commit history won't become cleaner :) I'll look into it probably this or next week or something

File: core/src/com/unciv/ui/tutorials/TutorialController.kt (hunk 1 above)
Comment (xlenstra) [reply core/src/com/unciv/ui/tutorials/TutorialController.kt:86]: > Nicely done. Wasn't there another virtually generated page in there, to compare....? No, that was Religion, wasn't it? 

I can't think of one, and the quick look I cast on the civilopedia code before writing this didn't reveal any other generated pages like this.

>And - "\n": Is that a FormattedLine collection or Strings? If the former, there is a "styled" empty line with a global-const controlled spacing (smaller than an entire line when I designed it) - if I remember correctly - thus mixing gives barely detectable line spacing inconsistencies.

I have no idea, to be honest, but the line breaks looked fine, and especially if its 'barely detectable', I don't think anyone will notice or mind it.

Comment (SomeTroglodyte) [reply core/src/com/unciv/ui/tutorials/TutorialController.kt:86]: OK, `lines` in there is strings, a compromise back then, so `\n` is fine

**Comment by yairm210** [comment]
"we love the king day" is another 'conditional bonus' that could enter the list :)
I'm not saying to implement all of these for the first PR, that would make it rather large
But I am saying that conditional uniques happen in many places in a way that's currently hardcoded

**Comment by xlenstra** [comment]
Should be ready now. Only thing that I'm still unsure about is if and how this information should be provided in the civilopedia.

File: android/assets/jsons/Civ V - Gods & Kings/GlobalUniques.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
+{
+    "name": "Global uniques", 
+    "uniques": [
+        "[-75]% growth [in all cities] <when between [-10] and [0] Happiness>",
+        "Nullifies [Food] [in all cities] <when below [-10] Happiness>",
```
Comment (yairm210) [inline android/assets/jsons/Civ V - Gods & Kings/GlobalUniques.json:5]: Nullifies Growth, not food, no? Nullifies food means everyone starves. And THAT means that we can just use the "[-100]% growth [in all cities] <when below [-10] Happiness>"

**Comment by yairm210** [comment]
Looks really good!

File: android/assets/jsons/Civ V - Gods & Kings/GlobalUniques.json (hunk 2 above)
Comment (xlenstra) [reply android/assets/jsons/Civ V - Gods & Kings/GlobalUniques.json:5]: That was my first plan as well, but then I discovered that if you have multiple growth bonuses (such as from landed elite), the other ones still provide food, which is not what you want. I will change it to nullify growth, though

**Review by yairm210 — APPROVED** [review]

File: core/src/com/unciv/ui/tutorials/TutorialController.kt [hunk 3]
```diff
@@ -70,10 +71,13 @@ class TutorialController(screen: BaseScreen) {
...
+        val civilopediaTutorials = tutorials.filter {
             Tutorial.findByName(it.key)!!.isCivilopedia
-        }.map {
-            CivilopediaTutorial(it.key, it.value)
+        }.map { tutorial ->
+            val lines = tutorial.value
+            CivilopediaTutorial(tutorial.key, lines)
         }
+        return civilopediaTutorials
+    }
```
Comment (will-ca) [inline core/src/com/unciv/ui/tutorials/TutorialController.kt:82]: Is the `ruleset` parameter not actually used anywhere?

**Review by will-ca — APPROVED** [review]

File: android/assets/jsons/Civ V - Gods & Kings/GlobalUniques.json [hunk 4]
```diff
@@ -0,0 +1,25 @@
+{
+    "name": "Global uniques",
```
[the remark this answers is not in the document]
Comment (xlenstra) [reply android/assets/jsons/Civ V - Gods & Kings/GlobalUniques.json:2]: I'd implement that with a unique for modOptions

File: core/src/com/unciv/logic/city/CityStats.kt [hunk 5]
```diff
@@ -294,14 +310,14 @@ class CityStats(val cityInfo: CityInfo) {
...
-            } else if (currentConstruction is BaseUnit) {
-                cityInfo.getMatchingUniques(UniqueType.PercentProductionUnits)
-            } else { // Science/Gold production
-                sequenceOf()
+            when {
+                currentConstruction is BaseUnit ->
+                    cityInfo.getMatchingUniques(UniqueType.PercentProductionUnits)
+                currentConstruction is Building && currentConstruction.isAnyWonder() ->
+                    cityInfo.getMatchingUniques(UniqueType.PercentProductionWonders)
+                currentConstruction is Building && !currentConstruction.isAnyWonder() ->
```
Comment (SomeTroglodyte) [inline core/src/com/unciv/logic/city/CityStats.kt:318]: second `isAnyWonder` call is now redundant, though good for readability.

File: core/src/com/unciv/models/ruleset/unique/UniqueType.kt [hunk 6]
```diff
@@ -95,7 +96,8 @@ enum class UniqueType(val text: String, vararg targets: UniqueTarget, val flags:
     AllStatsSignedPercentFromObject("+[amount]% yield from every [tileFilter]", UniqueTarget.Global, UniqueTarget.FollowerBelief),
     StatPercentFromReligionFollowers("[amount]% [stat] from every follower, up to [amount]%", UniqueTarget.FollowerBelief),
     BonusStatsFromCityStates("[amount]% [stat] from City-States", UniqueTarget.Global),
-    StatPercentBonusCities("[amount]% [stat] [cityFilter]", UniqueTarget.Global, UniqueTarget.FollowerBelief),
+    NullifiesStat("Nullifies [stat] [cityFilter]", UniqueTarget.Global),
+    NullifiesGrwoth("Nullifies Growth [cityFilter]", UniqueTarget.Global),
```
Comment (SomeTroglodyte) [inline core/src/com/unciv/models/ruleset/unique/UniqueType.kt:100]: Nobody saw this typo? Sounds cool though.

Comment (will-ca) [reply core/src/com/unciv/models/ruleset/unique/UniqueType.kt:100]: …I already had four comments, and figured it'd be easy to change anyway. 🤷

File: core/src/com/unciv/logic/city/CityStats.kt [hunk 7]
```diff
@@ -592,15 +624,12 @@ class CityStats(val cityInfo: CityInfo) {
...
-            if (happiness < 0) {
-                // Unhappiness -75% to -100%
-                val foodReducedByUnhappiness = if (happiness <= -10) Stats(food = totalFood * -1)
-                    else Stats(food = (totalFood * -3) / 4)
-                newFinalStatList.add("Unhappiness", foodReducedByUnhappiness)
-            } else if (cityInfo.isWeLoveTheKingDay()) {
+            val growthBonuses = getGrowthBonus(totalFood)
+            renameStatmapKeys(growthBonuses)
+            for (growthBonus in growthBonuses) {
+                newFinalStatList.add("${growthBonus.key} (Growth)", growthBonus.value)
```
[the remark this answers is not in the document]
Comment (xlenstra) [reply core/src/com/unciv/logic/city/CityStats.kt:630]: That is a good point actually

File: core/src/com/unciv/models/ruleset/unique/UniqueType.kt (hunk 6 above)
Comment (xlenstra) [reply core/src/com/unciv/models/ruleset/unique/UniqueType.kt:100]: I'll make a follow-up PR adressing the points of these comments later today :)

**Pull request merged**